components/status_dropdown/status_dropdown.tsx

```diff
@@ -63,35 +69,99 @@ export default class StatusDropdown extends React.PureComponent <Props, State> {
+            width: 0,
+        };
+    }
+
+    getCurrentDateTime = (tz: UserTimezone, enable: boolean): Date => {
+        let currentDate = new Date();
+        if (enable) {
+            if (tz.useAutomaticTimezone) {
+                currentDate = getCurrentDateTimeForTimezone(tz.automaticTimezone);
+            } else {
+                currentDate = getCurrentDateTimeForTimezone(tz.manualTimezone);
+            }
+        }
+        return currentDate;
     }
 
-    setStatus = (status: string): void => {
+    setStatus = (status: string, dndEndTime: any): void => {
         this.props.actions.setStatus({
             user_id: this.props.userId,
             status,
+            dnd_end_time: dndEndTime,
         });
     }
 
+    isUserOutOfOffice = (): boolean => {
+        return this.props.status === UserStatuses.OUT_OF_OFFICE;
+    }
+
     setOnline = (event: Event): void => {
         event.preventDefault();
-        this.setStatus(UserStatuses.ONLINE);
+        this.setStatus(UserStatuses.ONLINE, '');
     }
 
     setOffline = (event: Event): void => {
         event.preventDefault();
-        this.setStatus(UserStatuses.OFFLINE);
+        this.setStatus(UserStatuses.OFFLINE, '');
     }
 
     setAway = (event: Event): void => {
         event.preventDefault();
-        this.setStatus(UserStatuses.AWAY);
+        this.setStatus(UserStatuses.AWAY, '');
     }
 
-    setDnd = (event: Event): void => {
+    setDndUntimed = (event: Event): void => {
         event.preventDefault();
-        this.setStatus(UserStatuses.DND);
+        this.setStatus(UserStatuses.DND, '');
+    }
+
+    setDnd = (index: number): void => {
+        const currentDate = this.getCurrentDateTime(this.props.userTimezone, this.props.isTimezoneEnabled);
+        const currentTime = currentDate.getTime();
+        let endTime = new Date(currentTime);
+        switch (index) {
+        case 0:
+            // add 30 minutes in current time
+            endTime = new Date(currentTime + (30 * 60 * 1000));
+            break;
+        case 1:
+            // add 1 hour in current time
+            endTime = new Date(currentTime + (1 * 60 * 60 * 1000));
+            break;
+        case 2:
+            // add 2 hours in current time
+            endTime = new Date(currentTime + (2 * 60 * 60 * 1000));
+            break;
+        case 3:
+            // add one day in current date and set hours to last minute of the day
+            endTime = new Date(currentDate);
+            endTime.setDate(currentDate.getDate() + 1);
+            endTime.setHours(23, 59, 59, 999);
+            break;
+        }
+
+        const dndEndTime = toUTCUnix(endTime);
+        this.setStatus(UserStatuses.DND, dndEndTime);
+    }
+
+    setCustomTimedDnd = (): void => {
+        const dndCustomTimePicker = {
+            modalId: ModalIdentifiers.DND_CUSTOM_TIME_PICKER,
+            dialogType: DndCustomTimePicker,
+            dialogProps: {
+                userId: this.props.userId,
```

Comment:
I don't think we need this, we already import this field in via the redux connector.